000X-B-C-I-R-S-error-handling-codes.md

```diff
@@ -0,0 +1,337 @@
+  // is typically the registered service name of the tool or product that
+  // generates the error. Example: "pubsub.googleapis.com". If the error is
+  // generated by some common infrastructure, the error domain must be a
+  // globally unique value that identifies the infrastructure. For Google API
+  // infrastructure, the error domain is "googleapis.com".
+  string domain = 2;
+
+  // Additional structured details about this error.
+  //
+  // Keys should match /[a-zA-Z0-9-_]/ and be limited to 64 characters in
+  // length. When identifying the current value of an exceeded limit, the units
+  // should be contained in the key, not the value.  For example, rather than
+  // {"instanceLimit": "100/request"}, should be returned as,
+  // {"instanceLimitPerRequest": "100"}, if the client exceeds the number of
+  // instances that can be created in a single (batch) request.
+  map<string, string> metadata = 3;
+}
+```
+
+
+Example error:
+```json
+{
+  "error": {
+    "code": 400,
+    "message": "API key not valid. Please pass a valid API key.",
+    "status": "INVALID_ARGUMENT",
+    "details": [
+      {
+        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+        "reason": "API_KEY_INVALID",
+        "domain": "googleapis.com",
+        "metadata": {
+          "service": "translate.googleapis.com"
+        }
+      }
+    ]
+  }
+}
+```
+
+
+Below is partial table of the Standard Error code provided by gRPC and how they map to HTTP error codes. The entire list can found in the following links:
+- [gRPC Codes Table]https://cloud.google.com/apis/design/errors#handling_errors
+- (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]
+
+
+|HTTP | gRPC      | Description |
+|---- | ----------- | ----------- |
+|200  |	OK		  | No error. |
+|400  |	INVALID_ARGUMENT |	Client specified an invalid argument. Check error message and error details for more information. |
+|400  |	FAILED_PRECONDITION |	Request can not be executed in the current system state, such as deleting a non-empty directory. |
+|400  |	OUT_OF_RANGE |	Client specified an invalid range. |
+|401  |	UNAUTHENTICATED	| Request not authenticated due to missing, invalid, or expired OAuth token. |
+|403  |	PERMISSION_DENIED |	Client does not have sufficient permission. This can happen because the OAuth token does not have the right scopes, the client doesn't have permission, or the API has not been enabled. |
+|404  |	NOT_FOUND |	A specified resource is not found. |
+|409  |	ABORTED	| Concurrency conflict, such as read-modify-write conflict. |
+
+### Error Details Classification
+The following tables shows the propsosed error codes used in the **reason** for the **google.rpc.ErrorInfo** within the various levels of Dapr modules:
+
+**INIT**
+| Dapr Module | Description |
+| ----------- | ----------- |
+| CLI         | DPR_INIT_CLI_[1-1000] |
+| Self-hosted | DPR_INIT_SLF_[1-1000] |
```

Comment:
Unless we really need the space, can we just spell it out? If we need to save space, we can truncate the message.